src/shared/modules/auth/common.js

```diff
@@ -0,0 +1,217 @@
+  }
+
+  discoveredSSOProviders.forEach(provider => {
+    if (!provider) return
+    if (
+      !mandatoryKeysForSSOProviders.every(key => provider.hasOwnProperty(key))
+    ) {
+      authLog(
+        `Dropping invalid discovered SSO provider with id: "${provider.id}", missing key`
+      )
+      return
+    }
+    if (
+      !mandatoryKeysForSSOProviderParams.every(key =>
+        provider?.params?.hasOwnProperty(key)
+      )
+    ) {
+      authLog(
+        `Dropping invalid discovered SSO provider with id: "${provider.id}", missing params key`
+      )
+      return
+    }
+    if (
+      currentSSOProviders.find(crntProvider => crntProvider.id === provider.id)
+    ) {
+      if (isLocalhostOrigin) {
+        const idx = currentSSOProviders.findIndex(
+          crntProvider => crntProvider.id === provider.id
+        )
+        currentSSOProviders.splice(idx, 1)
+        authLog(`Updating SSO provider with id: "${provider.id}"`)
+      } else {
+        authLog(
+          `Not accepting discovered SSO provider with id: "${provider.id}", id exists already`
+        )
+        return
+      }
+    }
+    currentSSOProviders.push(provider)
+  })
+
+  window.sessionStorage.setItem(
+    AUTH_STORAGE_SSO_PROVIDERS,
+    JSON.stringify(currentSSOProviders)
+  )
+  authLog('Checked and merged SSO providers')
+}
+
+export const getSSOProvidersFromStorage = () => {
+  const ssoProviders = JSON.parse(
+    window.sessionStorage.getItem(AUTH_STORAGE_SSO_PROVIDERS)
+  )
+  if (!ssoProviders || !ssoProviders.length) {
+    authLog('No SSO providers in (local) storage found')
+    return []
+  }
+  if (!window.isSecureContext) {
+    authLog(
+      'This application is NOT executed in a secure context. SSO support is therefore disabled. Load the application in a secure context to proceed with SSO.',
+      'warn'
+    )
+    return []
+  }
+  return ssoProviders
+}
+
+export const getSSOProviderByIdpId = idpId => {
+  const ssoProviders = getSSOProvidersFromStorage()
+
+  const selectedSSOProvider = ssoProviders.find(
+    provider => provider.id === idpId
+  )
+  if (!selectedSSOProvider) {
+    authLog(`No matching SSO provider to passed in argument: ${idpId}`, 'warn')
+    return null
+  }
+  return selectedSSOProvider
+}
+
+export const getCredentialsFromAuthResult = (result, idpId) => {
+  const emptyCredentials = { username: '', password: '' }
+  authLog(`Attempting to assemble credentials for idp_id: ${idpId}`)
+
+  if (!result || !idpId) {
+    authLog('No result or idp_id passed in', 'warn')
+    return emptyCredentials
+  }
+
+  const selectedSSOProvider = getSSOProviderByIdpId(idpId)
+  if (!selectedSSOProvider) return emptyCredentials
+
+  const tokenTypePrincipal =
+    selectedSSOProvider.config?.['token_type_principal'] ||
+    defaultTokenTypePrincipal
+  authLog(
+    `Credentials, using token type "${tokenTypePrincipal}" to retrieve principal`
+  )
+
+  const parsedJWT = jwtDecode(result[tokenTypePrincipal])
```

Comment:
We can add typing based on this: https://github.com/auth0/jwt-decode#use-with-typescript

Reply:
Also we should catch the `InvalidTokenError` here...